plugins/modules/ec2_dedicated_host.py

```diff
@@ -0,0 +1,479 @@
+    host_info = None
+    hosts = describe_ec2_dedicated_hosts(client, HostIds=[host_id])
+    if hosts:
+        host_info = hosts[0]
+    return host_info
+
+
+def tags_match(match_tags: Dict[str, str], candidate_tags: Dict[str, str]) -> bool:
+    return all((k in candidate_tags and candidate_tags[k] == v for k, v in match_tags.items()))
+
+
+def get_ec2_dedicated_host(client, module: AnsibleAWSModule) -> Optional[Dict[str, Any]]:
+    host_info = None
+    # Select all states except 'released' and 'released-permanent-failure'
+    params = {"Filter": [{"Name": "state", "Values": ["available", "under-assessment", "permanent-failure"]}]}
+
+    lookup = module.params.get("lookup")
+    host_id = module.params.get("host_id")
+    ansible_tags = module.params.get("tags")
+
+    if lookup == "host_id":
+        params["HostIds"] = [host_id]
+    elif lookup == "tag" and not ansible_tags:
+        # lookup == 'tag' but tags was not provided
+        return host_info
+    hosts = describe_ec2_dedicated_hosts(client, **params)
+    if hosts:
+        if lookup == "tag":
+            count = 0
+            for h in hosts:
+                aws_tags = boto3_tag_list_to_ansible_dict(h.get("Tags", {}))
+                if aws_tags == ansible_tags:
+                    host_info = h
+                    count += 1
+            if count > 1:
+                module.fail_json(msg=f"Tags provided do not identify a unique dedicated host ({count} found).")
+        else:
+            host_info = hosts[0]
+
+    return host_info
+
+
+def format_output(host: Dict[str, Any]) -> List[Dict[str, Any]]:
+    host["tags"] = boto3_tag_list_to_ansible_dict(host.pop("Tags", {}))
+    return camel_dict_to_snake_dict(host, ignore_list=["tags"])
+
+
+def release_host(client, module: AnsibleAWSModule, existing: Dict[str, Any]) -> NoReturn:
+    if not existing:
+        module.exit_json(changed=False)
+    host_id = module.params.get("host_id")
+    if module.check_mode:
+        module.exit_json(changed=True, msg=f"Would have released dedicated host '{host_id}' if not in check mode.")
+    changed = release_ec2_dedicated_host(client, host_id)
+    module.exit_json(changed=changed, host=format_output(existing))
+
+
+def create_or_update_host(client, module: AnsibleAWSModule, existing: Dict[str, Any]) -> NoReturn:
+    host_id = module.params.get("host_id")
+    changed = False
+    if existing:
+        # Update dedicated host
+        # Check if one of the following has changed ('HostRecovery', 'InstanceType', 'InstanceFamily', 'HostMaintenance', 'AutoPlacement')
+        host_maintenance = module.params.get("host_maintenance")
+        instance_type = module.params.get("instance_type")
+        instance_family = module.params.get("instance_family")
+        auto_placement = module.params.get("auto_placement")
+        host_recovery = module.params.get("host_recovery")
+        params_to_update = {}
+        if host_recovery != existing["HostRecovery"]:
```

Comment:
It is assumed that `existing` dict has all the keys. Is the assumption always true? Can we use .get() to prevent errors?